Fix mixed LR

akva2 opened the pull request:
Some fixes for the mixed LR ASM module. I can now run 'mixed' Stokes fine through LR. More stuff will be amended to fix compatible stokes + LR, so hold on the merge (but feel free to review now)!

@kmokstad please have a look at the printing commit.

akva2 commented:
there we go. compatible was easier than anticipated. runs fine as well now.

kmokstad commented on src/ASM/LR/ASMu2Dmx.C:
```diff
@@ -668,3 +668,22 @@ bool ASMu2Dmx::refine (const LR::RefineData& prm,
+  if (!basis) {
+    std::cerr << "Asked to get coordinate for node " << inod
+              << ", but only have " << this->getBasis(b)->nBasisFunctions() << std::endl;
+    return Vec3();
```
Somewhat unclear error message, consider adding at the end: " nodes in basis "<< b+1

akva2 replied:
e75dd4a

kmokstad commented on src/ASM/LR/ASMu2Dmx.C:
```diff
@@ -668,3 +668,22 @@ bool ASMu2Dmx::refine (const LR::RefineData& prm,
+  }
+  return Vec3(&(*basis->cp()),nsd);
+
+}
```
Remove blank line 688

kmokstad commented on src/ASM/FiniteElement.C:
```diff
@@ -42,7 +48,7 @@ std::ostream& FiniteElement::write (std::ostream& os) const
   this->FiniteElement::write(os);
-  for (size_t b = 0; b+1 < Nx.size(); b++)
+  for (size_t b = 0; b < Nx.size(); b++)
   {
```
Not sure, but to me it looks like it should have been "for (size_t b = 1; b < Nx.size(); b++)"
Or else the first basis is printed twice, due to calling the parent class method?

akva2 replied:
first basis is not in Nx. Nx is only the additional bases (see MxFiniteElement constructor). it's this way due to inheritance stuff (we want to look like a FiniteElement for the first basis).

kmokstad replied:
OK, just disregard my previous comment, your suggested fix is correct. (I did not see the rest of the body, which you need to do in order to see that)

kmokstad commented on src/ASM/FiniteElement.h:
```diff
@@ -114,6 +114,8 @@ class MxFiniteElement : public FiniteElement
   //! \brief Writes the finite element object to the given output stream.
   virtual std::ostream& write(std::ostream& os) const;
+  //! \brief Global Output stream operator.
+  friend std::ostream& operator<<(std::ostream& os, const MxFiniteElement& fe);
```
Do you really need to overload this operator? Since it only invokes FiniteElement::write which is virtual, the correct behaviour will be achieved without it when invoked with a MxFiniteElement object as argument value.

akva2 replied:
yes, i need the implementation. it prints as a FiniteElement otherwise. just try for yourself.

akva2 replied:
Ja jeg var like overrasket selv, men det var det debuggeren fortalte meg.

kmokstad replied:
The conclusion on this was that it is _not_ needed, @akva2 probably triggered a bug in the compiler, or something. I just pushed the one-liner fix (line 45) plus a test-program, such that this commit can now be removed from this PR.

kmokstad commented on src/ASM/FiniteElement.C:
```diff
@@ -20,6 +20,12 @@ std::ostream& operator<< (std::ostream& os, const FiniteElement& fe)
+  return fe.write(os);
+}
+
+
```
See comment below on the header file

VikingScientist commented:
Do we have a basic mixed LR test to accompany this PR?

akva2 commented:
rebased as the printing was handled separately. added some fixes, we can now run adaptive stokes with taylor-hood, div-compatible and, with some other fixes that will come separately, stabilized formulations.

akva2 commented:
as for the tests, i will setup tests, but i cannot yet as this pr depends on https://github.com/OPM/IFEM/pull/73

VikingScientist commented on src/ASM/LR/ASMu2Dmx.C:
```diff
@@ -633,14 +646,22 @@ bool ASMu2Dmx::refine (const LR::RefineData& prm,
+        int mult = 1;
+        if (ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS1 ||
+            ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS2) {
+          int p = m_basis[bas]->order(line->span_u_line_ ? 0 : 1)-1;
```
Actually, if the line is spanning the u-direction, then the reduced continuity is *across* it, i.e. in the v-direction. You will have to switch the polynomial degree you are looking up here.

VikingScientist commented on src/ASM/LR/ASMu2Dmx.C:
```diff
@@ -633,14 +646,22 @@ bool ASMu2Dmx::refine (const LR::RefineData& prm,
+            ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS2) {
+          int p = m_basis[bas]->order(line->span_u_line_ ? 0 : 1)-1;
+          int k = p - line->multiplicity_;
+          int q = m_basis[j]->order(line->span_u_line_ ? 0 : 1)-1;
```
The same here
